### PR TITLE
Ability to remove minetest.after once set

### PR DESCRIPTION
**Pull request by tenplus1**
This change allows a mod to remove any minetest.after once it has been set e.g.
```
local job = minetest.after(10.0, function()
    print("--- this will never be displayed")
end)

minetest.after(5.0, function()
    print("--- this one will")
end)

if job then
   job:cancel()
end

### CODE REVIEW

**Comment by rubenwardy** [comment]
The index will change when a job expires:

```lua
local jobs_l = #jobs
jobs[i] = jobs[jobs_l]
jobs[jobs_l] = nil
```

**Comment by tenplus1** [comment]
Good point, how about replacing function with an empty, would that be better:
```
function core.remove_after(num)
	if jobs[num] then
		jobs[num].func = function() end
	end
end

**Comment by rubenwardy** [comment]
The problem is that num may be incorrect after a job expires or a job is removed

**Comment by tenplus1** [comment]
We'd need some way to verify it's still the same minetest.after that's being removed... hrm...

**Comment by rubenwardy** [comment]
You could have another file-local table that has constant ids, and then overwrite the function

**Comment by tenplus1** [comment]
Wouldn't that over-complicate things ?

**Comment by rubenwardy** [comment]
It'll make it work, this PR currently doesn't

**Comment by tenplus1** [comment]
Thanks for the input rubenwardy, will tinker tomorrow and see what I come up with :)

**Comment by sorcerykid** [comment]
My Job Control mod provides a pretty straightforward object-oriented approach to managing minetest.after timers. Instead of cancelling timers globally by ID, the timers themselves have a `cancel()` method.

**Comment by Desour** [comment]
Can't you just return the new entry in `jobs` and then set the function to an empty one in `remove_after`:
```lua
function core.after(after, func, ...)
	assert(tonumber(after) and type(func) == "function",
		"Invalid minetest.after invocation")
	local expire = time + after
	local new_job = {
		func = func,
		expire = expire,
		arg = {...},
		mod_origin = core.get_last_run_mod()
	}
	jobs[#jobs + 1] = new_job
	time_next = math.min(time_next, expire)
	return new_job
end

function core.remove_after(job)
	job.func = function() end
end
```

**Comment by rubenwardy** [comment]
> Can't you just return the new entry in jobs and then set the function to an empty one in remove_after:

d'oh, that's much easier

**Comment by tenplus1** [comment]
@Desour - That is indeed a much easier method, thank you :) will apply now...

**Comment by Desour** [comment]
> You could even have a method like so:
> 
> ```lua
> -- ...
> 		cancel = function(self)
> 			self.func = function() end
> 		end
> -- ...
> ```

You could even apply some information-hiding:
```lua
return { cancel = function() new_job.func = function() end end }
```
I'm pretty sure that otherwise there will be someone who manipulates one of the other fields.

**Comment by tenplus1** [comment]
@Desour - so having result = minetest.after(1.0, function() - whatever) would contain the cancelling function inside 'result' ? can you explain please

**Comment by Desour** [comment]
>so having result = minetest.after(1.0, function() - whatever) would contain the cancelling function inside 'result' ? can you explain please

It allows to simply call the cancel function on the return value, instead of passing the return value to another function.
And you can not access the new job table (the job is captured in the function).
Example:
```lua
local handle = minetest.after(0, print, "bla")
handle.arg[1] = "foo" -- not possible
handle.cancel() -- instead of minetest.stop_after(handle)
```

Another alternative would be returning the cancel function directly (but that might be ugly and doesn't allow to add further stuff to the handle in the future).

If you don't like calling a function on a return value, you could also return a new empty table and use this table as a key in a weak table:
```lua
local cancel_table = {}
setmetatable(cancel_table, {__mode = "k"} -- keys are weak

function core.after(after, func, ...)
	-- ...
	local cancellation_key = {}
	cancel_table[cancellation_key] = new_job
	return cancellation_key
end

function core.remove_after(cancellation_key)
	cancel_table[cancellation_key].func = function() end
end
```
(See http://www.lua.org/manual/5.1/manual.html#2.10.2 for more information on weak tables.)

**Comment by tenplus1** [comment]
Ah kewl :) that works a lot better than having a cancel function and safer like stated above :)  If you don't mind I'll use that in this pull request instead :P

File: doc/lua_api.txt [hunk 1]
```diff
@@ -5193,10 +5193,14 @@ Sounds
 Timing
 ------
 
-* `minetest.after(time, func, ...)`
+* `minetest.after(time, func, ...)` : returns job table to use as below.
     * Call the function `func` after `time` seconds, may be fractional
     * Optional: Variable number of arguments that are passed to `func`
 
+* `job.cancel()`
```
Comment (rubenwardy) [inline doc/lua_api.txt:5200]: `job:cancel()`

File: doc/lua_api.txt [hunk 2]
```diff
@@ -5193,10 +5193,14 @@ Sounds
...
+    * Removes already registered minetest.after function and replaces
```
Comment (rubenwardy) [inline doc/lua_api.txt:5201]: This is an implementation detail

**Comment by tenplus1** [comment]
@rubenwardy - thanks, lua_api.txt updated, sorry about all the commits so far :P

**Comment by sorcerykid** [comment]
> @Desour - so having result = minetest.after(1.0, function() - whatever) would contain the cancelling function inside 'result' ? can you explain please

It is the same as whtat I suggested at the beginning of the thread: having a cancel method. Here is how I implemented it in my Job Control mod: https://bitbucket.org/sorcerykid/cronjob/src/master/init.lua

**Comment by tenplus1** [comment]
@sorcerykid - wow, your job control mod has a lot more included :)  I'm just happy we now have a way to cancel any minetest.after() safely :P

File: doc/lua_api.txt (hunk 2 above)
Comment (rubenwardy) [reply doc/lua_api.txt:5201]: By this, I mean that the user doesn't care how cancelling is implemented

> Cancels the function from being called, providing that it hasn't been called already

File: builtin/common/after.lua [hunk 3]
```diff
@@ -31,11 +31,16 @@ function core.after(after, func, ...)
...
 		"Invalid minetest.after invocation")
 	local expire = time + after
-	jobs[#jobs + 1] = {
+	local new_job = {
 		func = func,
 		expire = expire,
 		arg = {...},
-		mod_origin = core.get_last_run_mod()
+		mod_origin = core.get_last_run_mod(),
+		cancel = function(self)
```
Comment (SmallJoker) [inline builtin/common/after.lua:39]: This job table is not exposed to mods, hence this cancel function cannot be used -> superfluous.

**Review by SmallJoker — APPROVED** [review]
LGTM

**Comment by nerzhul** [comment]
whereas technically it's valid, which workflow do you get which needs this ?

**Comment by tenplus1** [comment]
@nerzhul - This could be used for many player effects when the player quits the server, or ambience mod to safely adjust sound gain without waiting for it to stop first and then reapply.

**Comment by nerzhul** [comment]
ok it's lgtm to me

**Review by nerzhul — APPROVED** [review]

**Pull request merged**